### PR TITLE
Save initial signup information for users to aid in spam prevention

### PR DESCRIPTION
**Pull request by techknowlogick**
This will allow instance admins to view signup pattern patterns for public instances. It is modelled after discourse, mastodon, and MediaWiki's approaches.

Note: This has privacy implications, but as the above-stated open-source projects take this approach, especially MediaWiki, which I have no doubt looked into this thoroughly, it is likely okay for us, too. However, I would be appreciative of any feedback on how this could be improved.

### CODE REVIEW

**Review by denyskon — CHANGES_REQUESTED** [review]
I support integrating this functionality, however, please make it opt-out so that instance admins can decide on their own if they do not want this feature due to privacy reasons.

**Review by denyskon — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -507,6 +507,9 @@ INTERNAL_TOKEN =
 ;; stemming from cached/logged plain-text API tokens.
 ;; In future releases, this will become the default behavior
 ;DISABLE_QUERY_AUTH_TOKEN = false
+;;
+;; On user registration, record the IP address and user agent of the user to help identify potential abuse.
+;; RECORD_INITIAL_IP = false
```
Comment (wxiaoguang) [inline custom/conf/app.example.ini]: The option name doesn't match its behavior, not only "IP"

File: models/user/user.go [hunk 2]
```diff
@@ -120,6 +120,11 @@ type User struct {
 	// (ex: in private Gitea instances user won't be allowed to see even organizations/repositories that are set as public)
 	IsRestricted bool `xorm:"NOT NULL DEFAULT false"`
 
+	// Store the initial registration of the user, to aid in spam prevention
+	// Ensure that one IP isn't creating many accounts (following mediawiki approach)
+	InitialIP        string `xorm:"-"`
+	InitialUserAgent string `xorm:"-"`
```
Comment (wxiaoguang) [inline models/user/user.go]: It shouldn't bloat the User model since they do not really belong to it.

File: models/user/user.go [hunk 3]
```diff
@@ -745,6 +750,22 @@ func createUser(ctx context.Context, u *User, createdByAdmin bool, overwriteDefa
...
+		if err = SetUserSetting(ctx, u.ID, "initial_ip", u.InitialIP); err != nil {
+			return err
+		}
+
+		// trim user agent string to a reasonable length, if necessary
+		userAgent := strings.TrimSpace(u.InitialUserAgent)
+		if len(userAgent) > 255 {
+			userAgent = userAgent[:255]
+		}
+		if err = SetUserSetting(ctx, u.ID, "initial_user_agent", userAgent); err != nil {
```
Comment (wxiaoguang) [inline models/user/user.go]: There are already consts, but not magic strings.

image

File: services/context/access_log.go [hunk 4]
```diff
@@ -86,6 +86,7 @@ func AccessLogger() func(http.Handler) http.Handler {
 				ResponseWriter: rw,
 				Ctx: map[string]any{
 					"RemoteAddr": req.RemoteAddr,
+					"UserAgent":  req.UserAgent,
```
Comment (wxiaoguang) [inline services/context/access_log.go]: Are you sure it is right? IIRC `req.UserAgent` is a function.

Comment (techknowlogick) [reply services/context/access_log.go]: This is code that came in from a different branch I'm working on. I've removed it.

File: models/user/user.go (hunk 2 above)
Comment (techknowlogick) [reply models/user/user.go]: I agree it shouldn't bloat it, but my alternative was to add an additional input to the `CreateUser` function, and so I chose this route. I can adjust it if you prefer. Perhaps I can create a `UserMeta` struct that could be passed and reused for other things, too?

**Comment by lafriks** [comment]
> I support integrating this functionality, however, please make it opt-out so that instance admins can decide on their own if they do not want this feature due to privacy reasons.

I don't really think there is a privacy problem here as  IP addresses should be kept for auditing purposes and browser user agent string is not person identifying information or anyhow related to privacy

File: models/user/user.go (hunk 2 above)
Comment (wxiaoguang) [reply models/user/user.go]: I think UserMeta is right, but I won't strongly require it or block this.

Comment (techknowlogick) [reply models/user/user.go]: refactored to use a struct.

**Comment by techknowlogick** [comment]
@lafriks, I'm inclined to merge this as is, with it disabled by default. Then, you could open a new PR to change it so a more in-depth discussion can be had.

**Review by delvh — APPROVED** [review]
I don't understand how this can be useful to detect spam/abuse.
But the implementation seems fine as it is.

**Comment by delvh** [comment]
What I do understand however, is that this is mostly intended for public instances.
So perhaps we should describe in the config docs that especially those instances should consider enabling it?

**Comment by lunny** [comment]
How about creating a `CreateUserOptions` as a parameter, `user` and others could be parameter of that struct.

**Comment by techknowlogick** [comment]
@delvh by recording signup IPs you can track if multiple users share the same address in case a spam ring creates many accounts. And the user agent is useful too as you can see additional behaviours

**Comment by techknowlogick** [comment]
@lunny yes, that can be done, but it's out of scope for this PR as I don't want to refactor too much in toys one.

**Comment by yp05327** [comment]
@lunny 
Should we move this PR to 1.23.0 milestone? As 1.23.0 is not released yet.

**Pull request merged**